core/src/impl/Kokkos_TeamMDPolicy.hpp

```diff
@@ -0,0 +1,374 @@
+                          : static_cast<TeamMDRangeLastNestLevel>(
+                                current_nest_level == -1),
+                      static_cast<TeamMDRangeParThread>(current_nest_level ==
+                                                        par_thread_nest_level),
+                      static_cast<TeamMDRangeParVector>(current_nest_level ==
+                                                        par_vector_nest_level)>;
+};
+
+// Structs to determine on which nested level parallelization happens.
+//   - Rank is Kokkos::Rank<TotalNestLevel, Iter>
+//     - TotalNestLevel is the total number of loop nests
+//     - Iter is whether to go forward or backward through ranks (i.e. the
+//       iteration order for MDRangePolicy)
+//   - ThreadAndVector determines whether both vector and thread parallelism is
+//   in use
+template <typename Rank, TeamMDRangeThreadAndVector ThreadAndVector>
+struct HostBasedNestLevel {
+  static constexpr bool is_direction_left =
+      (Rank::outer_direction == Iterate::Left);
+  static constexpr int par_rt  = is_direction_left ? Rank::rank - 1 : 0;
+  static constexpr int par_rv  = is_direction_left ? 0 : Rank::rank - 1;
+  static constexpr int invalid = -2;
+};
+
+template <typename Rank, TeamMDRangeThreadAndVector ThreadAndVector>
+struct AcceleratorBasedNestLevel {
+  static constexpr bool is_direction_left =
+      (Rank::outer_direction == Iterate::Left);
+
+  // If vector parallelism is in use, deploy thread parallelism on
+  // the second to the last nested level; otherwise, thread parallelism on the
+  // last nested level
+  static constexpr int left_par_rt =
+      (ThreadAndVector == TeamMDRangeThreadAndVector::Both) ? 1 : 0;
+
+  static constexpr int right_par_rt =
+      (ThreadAndVector == TeamMDRangeThreadAndVector::Both) ? Rank::rank - 2
+                                                            : Rank::rank - 1;
+
+  static constexpr int par_rt = is_direction_left ? left_par_rt : right_par_rt;
+
+  // Vector parallelism will always be on the last index
+  static constexpr int par_rv  = is_direction_left ? 0 : Rank::rank - 1;
+  static constexpr int invalid = -2;
+};
+
+// ThreadAndVectorNestLevel determines on which nested level parallelization
+// happens.
+//   - Rank is Kokkos::Rank<TotalNestLevel, Iter>
+//     - TotalNestLevel is the total number of loop nests
+//     - Iter is whether to go forward or backward through ranks (i.e. the
+//       iteration order for MDRangePolicy)
+//   - ThreadAndVector determines whether both vector and thread parallelism is
+//   in use
+template <typename Rank, typename ExecSpace,
+          TeamMDRangeThreadAndVector ThreadAndVector>
+struct ThreadAndVectorNestLevel;
+
+#ifdef KOKKOS_ENABLE_SERIAL
+template <typename Rank, TeamMDRangeThreadAndVector ThreadAndVector>
+struct ThreadAndVectorNestLevel<Rank, Serial, ThreadAndVector>
+    : HostBasedNestLevel<Rank, ThreadAndVector> {};
+#endif
```

Comment:
This should be in backend-specific files, probably together with other `MDRange` implementation files.

Reply:
Refactored in [ef6d137](https://github.com/kokkos/kokkos/pull/5238/commits/ef6d137d66aecc586bfe7b9ba673714a8978d40b).

Moved these structs to each backend's `Kokkos_[backend]_MDRangePolicy.hpp` file.